### PR TITLE
Linkify code spans in docs

### PR DESCRIPTION
**Pull request by ianprime0509**
By keeping around the `Scope`s built while walking a file and
associating them to their namespace declarations, it is possible to
resolve declarations by name. This, in turn, makes it simple to linkify
dotted identifier paths in code spans using the containing scope of the
associated declaration.

---

Memory usage comparison, measured using `wasm.instance.exports.memory.buffer.byteLength` after loading the root page (increase because the `Walk.Scope`s are persisted in memory now rather than being discarded after walking each file):

Before: 76,218,368
After: 79,429,632

Points of interest:

- `#std.fs` - contains quite a few doc links (including ones with more than one component)
- `#std.json` - contains doc links in the container doc comment
- `#std.debug.assert` - contains a link to another namespace (`std.testing`) (for some reason this link isn't detected by [current Autodoc](https://ziglang.org/documentation/master/std/#A;std:debug.assert))
- `#std.testing` - contains a few links (such as `math.approxEqAbs`) which should start being linkified correctly once [`field_access` is categorized as an alias](https://github.com/andrewrk/autodoc/blob/ddffab23d6ef03015260bc596270b15c128f3f26/src/Walk.zig#L134-L138)

### CODE REVIEW

**Comment by andrewrk** [comment]
I pushed some relevant commits that I expect to simplify this task.

**Comment by ianprime0509** [comment]
Thanks! The changes you made definitely made this implementation simpler and more reliable. Since the approach being taken here keeps the `Scope`s in memory (to allow name-based lookups), I included a Wasm memory usage comparison in the updated PR description.

Most of the links I could find when browsing around the docs are being resolved correctly. Of the ones that aren't, these are the categories I found as to why:

1. Not enclosed in code spans: it's certainly possible to linkify text outside `` `code spans` ``, but it would result in many false positive links if common words are used as identifiers (e.g. every occurrence of the word "path" in `std.fs` would be linked).
2. References an identifier that isn't defined in the current scope: for example, references like `std.io.Writer` in the `std.json` docs don't work because `json.zig` doesn't actually include the line `const std = @import("std");`, so `std` is not resolvable in that scope.
3. Not resolvable due to this TODO: https://github.com/andrewrk/autodoc/blob/ddffab23d6ef03015260bc596270b15c128f3f26/src/Walk.zig#L134-L138

**Comment by andrewrk** [comment]
I figured keeping the scopes might be necessary to solve this use case. Let me rejigger the memory layout with that in mind, I think if we make the scope tables long-lived then we can remove some other tables.

This would also be handy for rendering guides - they could be treated as markdown that is supposed to live in a certain scope (probably file level).

**Comment by andrewrk** [comment]
Actually I think this is a nice starting point - I'll merge this and then make some follow-up enhancements :+1:

File: src/main.zig [hunk 1]
```diff
@@ -463,6 +473,18 @@ fn render_docs(
     try renderer.render(parsed_doc, out.writer(gpa));
 }
 
+fn resolve_decl_path(decl_index: Decl.Index, path: []const u8) ?Decl.Index {
+    var path_components = std.mem.splitScalar(u8, path, '.');
+    var current_decl_index = decl_index.get().lookup(path_components.first()) orelse return null;
+    while (path_components.next()) |component| {
+        if (categorize_decl(current_decl_index, 0) == .alias) {
```
Comment (andrewrk) [inline src/main.zig:480]: please note 8d9f658ab8d1b547104b9e85aebd902da4b19f4c

**Comment by andrewrk** [comment]
> `#std.testing` - contains a few links (such as `math.approxEqAbs`) which should start being linkified correctly once [`field_access` is categorized as an alias](https://github.com/andrewrk/autodoc/blob/ddffab23d6ef03015260bc596270b15c128f3f26/src/Walk.zig#L134-L138)

check out bdb8f6dab7cb916c4c422070c0221995f1c70fc5 :smiley:

**Comment by ianprime0509** [comment]
> check out [bdb8f6d](https://github.com/andrewrk/autodoc/commit/bdb8f6dab7cb916c4c422070c0221995f1c70fc5) 😃

Awesome! This is great, definitely helps fill in the missing pieces of this implementation.

**Pull request merged**